### ISSUE TITLE
Machine Learning Batch Processing

### ISSUE THREAD

**Issue by JRose1** [danforthcenter/plantcv#162]
Question:

Is it possible to use the plantcv-pipeline.py script for multiprocessing of a directory of images at once for running the naive bayes classifier? The plantcv-train.py script already trains on a directory, but from all I see there's only documentation for actually running the classifier on one image at a time.

**Comment by nfahlgren** [comment]
@JRose1 it's definitely possible. I updated the [machine learning tutorial](http://plantcv.readthedocs.io/en/latest/machine_learning_tutorial/) with a second method and some instructions on using both in conjunction with `plantcv-pipeline.py`. Hopefully that helps, but let me know if anything is unclear.

**Comment by JRose1** [comment]
Ah thank you, definitely some good new info in the parallel-image processing section. Having looked more closely at the naive_bayes.py training script, I'm curious as to whether it simply rewrites any old PDF file from before or actually adds the current image directory to its training memory? i.e. Is running naive_bayes.py on training set 1 and then training set 2 the same as running it on both training set 1 and 2 in the same command?

**Comment by nfahlgren** [comment]
@JRose1 it doesn't combine the results across two sets unless you include both image directories together in the same command. So running it twice on two sets with the same output file would erase the first set of PDFs and would only record the second. I don't think it would be technically possible to combine a new set into a previous set given only the PDF file because the PDFs are only an approximation of the true color channel distributions of the first set.

**Comment by JRose1** [comment]
Ah, that would have been extremely useful to have read in the documentation.

Would it not be possible to make it load a previous PDF and adjust it to new data? I'm assuming for each training sample it's essentially multiplying it old state (It's prior assumption) by some new probability associated with features in the new training data.  This would be independent of any other information in the training data, as the only piece of stored information is the probability distribution for each class.

**Comment by JRose1** [comment]
I'm also finding the naive bayes classifier is ending up skewed seemingly due to an imbalance in the training sets for each class. i.e. there are far more background pixels than foreground pixels, so it gets more experience with background pixels. Do you have a favored way of making class training more independent of amount of each class represented in the training data?

**Comment by nfahlgren** [comment]
I would think the most accurate way to update the PDFs would be to have the pixel data for both sets. I suppose you could theoretically create PDFs for the new dataset and somehow average it with the old PDFs but there wouldn't be any guarantee that weighting them equally is a valid representation of the population statistics.

There shouldn't be any skew because the datasets are designed to be balanced. For the naive Bayes method using images and masks, we randomly sample 10% of the pixels from the foreground (plant) and then randomly sample from the background the same number of pixels. So in the end both the foreground and background classes have the same sample size. For the naive Bayes multiclass method, the input is a table with one column per class. Since the table has to have the same number of rows for each class, the pixel samples have to be equally sized for this method as well.

**Comment by JRose1** [comment]
Hmm, but if there ends up being an imbalance in the test set (i.e. there is generally more background than foreground) would that affect its tendency to misclassify? As in, if you're automatically sampling 50% background and 50% foreground in training, does that not make it assume there will be about 50% of each in its classifications?

Also, what happens if there are not the same number of training values for each class when using multiclass pixel sampling? The script seems to run without errors and create a PDF, though with radically different quantities (I used only 2 rows for 255,255,255 = white background and over 150 for foreground) it seems to calculate infinity for the background probabilities.

**Comment by nfahlgren** [comment]
The classifier only learns what each class "looks" like in color space so it does not make any assumptions about the relative abundance of each class.

Hmm, I see. I think you are right. My intention was that the table should have the same number of rows for each class (which is not necessarily a good thing or convenient) but the way I wrote the code actually does not look like it will have an error if the number of rows are not equal. That might actually be a good thing in some cases where one class is relatively small or very uniform and collecting a thousand samples just does not make sense (e.g. a very white background). That being said, I would want to convince myself that the pixels were going into the correct class (it should as long as there are the correct number of tabs for each row).

The problem you are talking about with your white background example though has to do with the sample size. The kernel density estimator algorithm needs some minimal set of data with some variance to accurately generate a proper PDF. I don't have a good sense for what the strict minimum might be but I think you likely need somewhere in the range of 20-100 pixels and I think they need to have some (minimal) variation (i.e. not all 255,255,255). We have been discussing ways to incorporate uniform distributions (like solid white or black backgrounds) or even exclude the background as a specific class altogether, but we don't have that worked out all the way yet.
